### DISCUSSION TITLE
Testing: assertViewHas support for comparing eloquent collections

### DISCUSSION THREAD

**Discussion by fuel-wlightning** [livewire/livewire#4021] [3. Feature ideas]
Currently you cannot do:
```
public function test_can_filter_to_active_people()
    {
        $user = $this->_getUserWithTeamRole('admin');
        $activePeople = Person::factory()->count(2)->create([
            'group_id' => $user->group_id,
            'active' => 1,
        ]);
        Person::factory()->create([
            'group_id' => $user->group_id,
            'active' => 0,
        ]);

        $this->actingAs($user)->withSession(['group_id' => $user->group_id]);

        Livewire::test('admin.people.person-admin')
            ->assertViewHas('people', $activePeople);
    }
```
This is caused by the deep comparison that PHPUnit is doing internally, which triggers on various parts that are not really important (like "is this recently created"). If you do an assertViewHas on a single model, it ignores those parts because assertViewHas has a special case handling for Models.

I propose we add special case handling to assertViewHas for EloquentCollections adding the following conditional to livewire `src/Testing/Concerns/MakeAssertions.php` method `assertViewHas` will achieve this:
```
} elseif ($value instanceof \Illuminate\Database\Eloquent\Collection) {
      PHPUnit::assertEmpty($value->diff($this->lastRenderedView->gatherData()[$key]));
}
```
This allows an Eloquent collection of models to be detected, and compared with the diff method.

**Comment by fuel-wlightning** [comment]
As a current workaround, with the current test, you can run `$activePeople->refresh()` which refetches all the models, making the objects deep match, which makes phpUnit happy, however incurs extra db calls.

This could be done on the assertion line as follows:
``` 
Livewire::test('admin.people.person-admin')
            ->assertViewHas('people', $activePeople->refresh());
```

**Comment by joshhanley** [comment]
@fuel-wlightning makes sense. PR's are welcome! 🙂
